robin/model/multimodal_encoder/open_clip.py

```diff
@@ -1,116 +1,67 @@
+        self.hidden_size = None # Place Holder get this value in load_model function
+        self.dtype = None # Place Holder get this value in train function
+        self.is_loaded = False
         if not delay_load:
             self.load_model()
-        else:
-            pass
-        
-        # self.hidden_size = 1152 if "SO400M" in self.vision_tower_name else 768  # TODO: this is a hack, we should read this from config.embed_dim
-
-        self.device = None
-        self.dtype = None
         
     def load_model(self):
-        name = self.vision_tower_name.split("/")[-1]
-        
-        if os.path.exists(self.vision_tower_name):
-
-            config = OPENCLIP_CONFIG_MAP[name] if name in OPENCLIP_CONFIG_MAP.keys() else name
-
-            self.vision_tower, self.image_processor = create_model_from_pretrained(config, pretrained=self.vision_tower_name+'/open_clip_pytorch_model.bin')
-        else:
-            self.vision_tower, self.image_processor = create_model_from_pretrained(self.vision_tower_name)
-        
-
+        assert "/" in self.vision_tower_name, """
+            vision tower name and pretrained must be split with / in open clip eg: ViT-B-16/laion2b_s34b_b88k"""
         
-        self.vision_tower = self.vision_tower.visual
-
-        if "timm" in self.vision_tower_name or name.lower() in TIMM_ON_OPENCLIP:
-            self.hidden_size = self.vision_tower.trunk.embed_dim
-        else:
-            # self.hidden_size = self.vision_tower.num_features
-            # self.hidden_size = self.vision_tower.output_dim
-            self.hidden_size = self.vision_tower.proj.data.shape[0]
-            self.vision_tower.pshape = self.vision_tower.proj.data.shape
-            self.vision_tower.proj = None
-            self.vision_tower.output_tokens = True
-        
-        # self.vision_tower.trunk.output_tokens = True
-        
-        self.vision_tower.requires_grad_(False)
+        model_name, pretrained = self.vision_tower_name.split("/")
+        open_clip, self.image_processor = create_model_from_pretrained(model_name, pretrained)
+        self.vision_tower = open_clip.visual
+
+        self.hidden_size = self.vision_tower.proj.shape[0] # find the dim before the final proj
+        self.vision_tower.output_tokens = True
+        self.vision_tower.proj = None # Avoid cls and image patch token dim mismatch
+        # self.vision_tower.requires_grad_(False) # TODO, freeze the vision tower?
         self.is_loaded = True
         
-        
-        
-        
     def feature_select(self, image_forward_outs):
-        
-        assert False, ("not implemented")
-        
-        image_features = image_forward_outs.hidden_states[self.select_layer]
+        cls_token, patch = image_forward_outs
         if self.select_feature == 'patch':
-            image_features = image_features[:, 1:]
+            image_features = patch
         elif self.select_feature == 'cls_patch':
```

Comment:
We plan to support cls only, and cls_patch. If we *Really* want we can support patch only also.